### PR TITLE
Server Side Rendering

### PR DESCRIPTION
**Pull request by datapimp**
This is still a WIP but I am submitting it because some people were interested in seeing an approach to server side rendering. 

Essentially the way it works is that the root container of the app is a `react-helmet` component that is responsible for rendering the entire HTML payload.  This component is used to render the page on the server side by using `react-router` match function to handle the express request object and render the given page.

Credit for this approach goes 100% to another boilerplate: https://github.com/DominicTobias/universal-react/

A few other changes have been made such as using webpack to compile and run the server code.  

I will clean this up and improve upon it over the weekend hopefully.
## TODO
- [ ] improve webpack server config and make server side hmr work 
- [ ] improve documentation 
- [ ] add tests
- [ ] usage example: dynamic request handler in production
- [ ] usage example: static site generator

### CODE REVIEW

**Comment by tzarger** [comment]
@datapimp Hi Jonathan, this is an excellent start. I was looking in to start working on this too and the first thing I was looking into was how to handle redux-saga on server side. Are you planning on working that into the ssr? Thanks for the great start on this, I look forward to see your refinements.

**Comment by datapimp** [comment]
My goal is definitely to have as little of difference as possible in the way you think about redux-saga, just not sure how to go about enforcing some of the constraints that you have to put on yourself to be able to do that.  

I added isomorphic-fetch for example.  Would love some suggestions or tips from people with battle scars from doing this already

**Comment by tzarger** [comment]
@datapimp Jonathan, since redux-saga is a bit of a bear to use server side, almost hack like at best, have you used or looked into https://github.com/makeomatic/redux-connect for loading up state server side? redux-saga is definitely nice, but not sure it is the best choice when doing a universal application.

**Comment by mxstbr** [comment]
`redux-connect` is great, but doesn't quite solve the same problem as `redux-saga` does, specifically tying together decoupled components. (see [this talk](https://vimeo.com/168648012) at 14:40 for the saga specific part, but I encourage you to watch it in its entirety)

I'm strongly convinced that `redux-saga` will shape the way we build frontend applications for a long time, and I would very much despise getting rid of it!

**Comment by justingreenberg** [comment]
hey guys, @datapimp solid work as usual. i haven't had a time to checkout locally and review but wanted to provide a resource and one quick thought before i head out for the weekend

redux saga ssr reference implementation:
https://gist.github.com/sompylasar/5e7157e451f4b7268def9ae1ce01edd4#file-serverrenderloop-js

this is a solid pattern for handling render loop. you will immediately notice that it is not trivial lol

this was also helpful in understanding, but keep in mind our routes are async and injected:
https://github.com/yelouafi/redux-saga/blob/master/examples/real-world/server.js

in our case, invoking sagas from router should mitigate some of the constraints associated with ssr but requires some consideration. for example, our current implementation will work, but it's not actually doing anything... eg, we would want to set a default username and dispatch a load repos action, either from componentWillMount or using put effect in root saga... we will also need to explicitly close the open take channel on success.. just thinking out loud here

one other quick thought on SSR—

getting client/server parity for asset name resolution is a serious challenge... ie

```js
import Banner from './banner-metal.jpg';
```

obviously this will not work in native node environment, and if it did the asset name wouldn't match the one generated by webpack, breaking checksum when mounting on client... canReuseMarkup throws

in my experience, **the simplest possible solution is to monkey `require`** and use webpack's `loader-utils` to manually interpolate the output name.... here's a gist: 
https://gist.github.com/justingreenberg/f1735abffbae3778014f8d14b599a2ef#file-require-patch-js

```js
```

this is a hack, but i have found it very reliable... alternatively, since you are propsing we use webpack to compile server code as well, we could leverage webpack itself to handle these assets during compilation... i've worked with various webpack plugins/hacks with mixed experiences:
https://github.com/halt-hammerzeit/webpack-isomorphic-tools
https://github.com/exponentjs/asset-module-webpack-plugin

**TL;DR: if we are using webpack to compile server, i recommend using https://github.com/jchip/isomorphic-loader**

totally did not mean to write so much, now i really gotta run.. let me know if you guys have any questions :) have a great weekend

**Comment by datapimp** [comment]
@justingreenberg thanks a lot, thanks for the advice, I will definitely read everything.

@mxstbr don't worry i feel just as strongly about redux-saga

**Comment by datapimp** [comment]
@justingreenberg  any thoughts on the following approach?

Essentially container components / routes would declare a static method 

```javascript
export class Container extends Component {
  static readyOnActions (dispatch) {
    return Promise.all([ ])
  }
 
  componentDidMount () {
    Container.readyOnActions(this.props.dispatch)
  }
}
```

The server renderer would have access to the route's getComponent, getChildRoutes, getComponents, etc and any of them which have this method would then resolve the various promises before returning the final markup.

I think this might eliminate the need for a render loop but I am still digging into that code you sent to get a better grasp.

**Comment by Stalso** [comment]
@datapimp If I understand approach with `static readyOnActions` correctly, then it is implemented in [MERN](https://github.com/Hashnode/mern-starter). You can look at it. But if you look at  [this](https://github.com/yelouafi/redux-saga/blob/master/examples/real-world/server.js) code, you will see, how to use Saga `END` effect for ssr. Maybe it is reasonable to use this approach? And also we can use `redial ` with Saga, like it is described here -  https://github.com/yelouafi/redux-saga/issues/293 .

I am also trying to implement ssr for this boilerplate, and the main challenge for me is require of `sanitize.css` in `\app\containers\App\index.js`. It always fails on server side. Maybe it is because this css in externals during server-side build, and server-side webpack loaders for css (I have tried to use many approaches, [fake-style-loader](https://github.com/dferber90/fake-style-loader), for example) do not work? I do not know. I ended up with ignoring this `css` on server side like this:

    if(__CLIENT__)
      require('sanitize.css/sanitize.css');

And I was happy, until I have installed [react-flexbox-grid](https://github.com/roylee0704/react-flexbox-grid) which internally requires `flexboxgrid.css` . Of course I can not ignore it. And everything was broken. And no loaders helped me to correctly load this css.

 For application styles my loaders look like this and work:

    'css-loader/locals?localIdentName=[local]__[path][name]__[hash:base64:5]&modules&importLoaders=1&sourceMap!postcss-loader'

I do not know the reason, but images are loaded correctly using the same loaders as on client side. 

Also I have looked at your code and do not understand, how will you debug server side with your approach, and why do not you use [webpack-externals-plugin](https://github.com/Morhaus/webpack-externals-plugin)?

**Comment by oyeanuj** [comment]
@datapimp Came across another server-side rendering with redux-saga attempt that might have some useful ideas: https://github.com/xkawi/react-universal-saga

**Comment by tzarger** [comment]
@oyeanuj Yes, I just saw that too. Looks like it might do the trick here. Some things it is lacking is code splitting, etc. Not sure if that would be an issue since sagas are injected... I really would like to see an option here for SSR, does not need to be baked in by default, but a configuration would be sweet. It seems that SSR movement here has slowed down, and hope it is not dead. @datapimp

**Comment by tzarger** [comment]
@datapimp @justingreenberg @Stalso Do you know if anyone is working on this?

**Comment by Stalso** [comment]
@tzarger I am currently working on this,  but I am doing something wrong.  Let more experienced people solve this problem.  As I said in previous post,  css it the main problem for me,  when I try to make webpack build of the app for server rendering

**Comment by oyeanuj** [comment]
@justingreenberg @datapimp Any thoughts on @Stalso's previous comment with the CSS as that seems a big blocker at the moment?

**Comment by kachkaev** [comment]
What is the current state of this PR? Feels like after this feature is merged into the boilerplate it'll leave all the competitors so far behind!

What are you guys using to achive SSR in your React/CSS-modules projects meanwhile?

**Comment by kachkaev** [comment]
This repository might be a useful source for some ideas:
https://github.com/reactGo/reactGo

Just found it via http://andrewhfarmer.com/starter-project/, a pretty nice little app for choosing a boilerplate. The list seems to be up to date (20 Aug 2016), but is not 100% complete (react-create-app is missing, for instance).

**Comment by scf4** [comment]
@Stalso Did you manage to solve your issue?

**Comment by davidwilde** [comment]
I found this article about isomorphic react isomorphic - https://reactjsnews.com/isomorphic-react-in-real-life

An application that I'm working on was inspired by this boilerplate but I wanted it to be isomorphic because SEO is particularly important. My own application basically ripped out all of the CSS module stuff because I couldn't solve it at the time - and time constraints yadda yadda yadda.

However possibly following this guide, could lead to an isomorphic version of this application.

**Comment by gihrig** [comment]
@davidwilde That looks like a great article, thanks for sharing!

Another step in the 'universal' app direction, [react-engine](https://github.com/paypal/react-engine) recently came to my attention. Looks like it may solve some of the 'pain' points discussed in the above article.

**Comment by someoneinatree** [comment]
I've been looking at approaches for this, and managed to pull together a half-working (i.e. JS for reduxified routes, no working CSS yet) SSR implementation on my react-boilerplate project (forked ~4 months ago). This pull request being the most helpful for what I ended up doing! (Besides realizing I had no clue what webpack was actually doing)

Can check it out here: https://github.com/opensessions/opensessions

The important files being `internals/webpack/webpack.server.babel.js`, `package.json` for the `npm run build:server` command, `server/src/server.js`, and any of the working pages `app/containers/{HomePage,SessionView,ListSessions}/index.js`.

**Comment by asood123** [comment]
Wanted to mention this post here: https://medium.com/open-collective/asking-for-a-new-feature-on-an-open-source-project-3305deba1356#.6ehgqgzhc.

**Comment by mxstbr** [comment]
Since this is very out of date and will highly likely not be merged, I'll close it for now. Thanks for the start, this is a great reference point @datapimp!

**Comment by scf4** [comment]
@asood that sounds great. @mxstbr can you confirm this is legit?

**Comment by mxstbr** [comment]
This is totally legit.

**Comment by tzarger** [comment]
@mxstbr @asood123 Can we drop a message in the Gitter channel to everyone to see if we can raise the $850 mentioned in the article? I think it would be a great opportunity to get that done after all this time and would be interested in donating as well.

**Comment by HTeker** [comment]
Will this be merged in the master branch any time soon?

**Comment by Jayiitb** [comment]
@mxstbr Are we working on this?

**Comment by gretzky** [comment]
This is currently benched for the time being.
